File: 1052-max-satisfied/solution.go
//自己寫出來的
package main

import "fmt"
func maxSatisfied(customers []int, grumpy []int, minutes int) int {
	
	basicSatisfied := 0
	for idx, val := range customers {
		if grumpy[idx] == 0{
			basicSatisfied += val
		}
	}

	addSatisfied := 0
	for i := 0; i < minutes; i++ {
		if grumpy[i] == 1{
			addSatisfied += customers[i]
		}
	}
	maxAddSatisfied := addSatisfied

	left := 0
	for right := minutes; right < len(customers); right++ {
		if grumpy[right] == 1{
			addSatisfied += customers[right]
		}
		if grumpy[left] == 1 {
			addSatisfied -= customers[left]
		}
		maxAddSatisfied = max(maxAddSatisfied, addSatisfied)
		left++
	}
	return basicSatisfied + maxAddSatisfied
}

func main(){
	customers := []int {4, 10, 10}
	grumpy :=	 []int {1,1,0}
	minutes := 2
	fmt.Println("Ans = ", maxSatisfied(customers, grumpy, minutes))
}
/*
執行方式:
go run ./1025-max-satisfied
*/

File: 1052-max-satisfied/Solution.java
class Solution1052 {
    public static int maxSatisfied(int[] customers, int[] grumpy, int minutes) {
        int basicSatisfied = 0;
        int i =0;
        for (int val : customers){
            if (grumpy[i] == 0)
                basicSatisfied += val;
            i++;
        }
        int addSatisfied = 0;
        for (int j = 0; j < minutes; j++)
            if (grumpy[j] == 1)
                addSatisfied += customers[j];
        int maxAddSatisifed = addSatisfied;

        int l =0;
        for (int r = minutes; r < customers.length; r++) {
            if (grumpy[l] == 1)
                addSatisfied -= customers[l];
            if (grumpy[r] == 1)
                addSatisfied += customers[r];
            System.out.println(addSatisfied);
            maxAddSatisifed = Math.max(maxAddSatisifed, addSatisfied);
            l++;
        }
        return basicSatisfied + maxAddSatisifed;
    }
    static void main(String[] args) {
        int[] customers = {1,0,1,2,1,1,7,5};
        int[] grumpy =    {0,1,0,1,0,1,0,1};
        System.out.println("Ans = " + maxSatisfied(customers, grumpy, 3));
    }
}